dart/dynamics/InverseKinematics.h

```diff
@@ -0,0 +1,841 @@
+  ///
+  /// Note that a target will automatically be created for the IK module upon
+  /// instantiation, so you typically do not need to use this function. If you
+  /// want to attach the target to an arbitrary (non-SimpleFrame) reference
+  /// frame, you can do getTarget()->setParentFrame(arbitraryFrame)
+  void setTarget(std::shared_ptr<SimpleFrame> _newTarget);
+
+  /// Get the target that is being used by this IK module. You never have to
+  /// check whether this is a nullptr, because it cannot ever be set to nullptr.
+  std::shared_ptr<SimpleFrame> getTarget();
+
+  /// Get the target that is being used by this IK module. You never have to
+  /// check whether this is a nullptr, because it cannot ever be set to nullptr.
+  std::shared_ptr<const SimpleFrame> getTarget() const;
+
+  /// Get the JacobianNode that this IK module operates on.
+  JacobianNode* getNode();
+
+  /// Get the JacobianNode that this IK module operates on.
+  const JacobianNode* getNode() const;
+
+  /// This is the same as getNode(). It is used by the InverseKinematicsPtr to
+  /// provide a common interface for the various IK smart pointer types.
+  JacobianNode* getAffiliation();
+
+  /// This is the same as getNode(). It is used by the InverseKinematicsPtr to
+  /// provide a common interface for the various IK smart pointer types.
+  const JacobianNode* getAffiliation() const;
+
+  /// Compute the Jacobian for this IK module's node, using the Skeleton's
+  /// current joint configuration and the DOFs that have been assigned to the
+  /// module.
+  const math::Jacobian& computeJacobian() const;
+
+  /// Get the current joint configuration of the Skeleton. This will only
+  /// include the DOFs that have been assigned to this IK module, and the
+  /// components of the vector will correspond to the components of getDofs().
+  Eigen::VectorXd getConfiguration() const;
+
+  /// Get the current joint configuration of the Skeleton. This must only
+  /// include the DOFs that have been assigned to this IK module, and the
+  /// components of the vector must correspond to the components of getDofs().
+  void setConfiguration(const Eigen::VectorXd& _q);
+
+  /// Clear the caches of this IK module. It should generally not be necessary
+  /// to call this function. However, if you have some non-standard external
+  /// dependency for your error and/or gradient method computations, then you
+  /// will need to tie this function to something that tracks changes in that
+  /// dependency.
+  void clearCaches();
+
+protected:
+
+  /// The InverseKinematics::Objective Function is simply used to merge the
+  /// objective and null space objective functions that are being held by an
+  /// InverseKinematics module. This class is not meant to be extended or
+  /// instantiated by a user. Call InverseKinematics::resetProblem() to set
+  /// the objective of the module's Problem to an InverseKinematics::Objective.
+  class Objective : public Function, public optimizer::Function
```

Comment:
I agree that it would probably be good to make it ``final``. I hadn't done so yet because I wasn't sure if some *might* have some reason to override it, but for now I think making it ``final`` is a good idea, and we can loosen that in the future if anyone has a need to override.